File: python/utils/ai_comment_generator.py
"""
AI Comment Generator for Social Media Engagement

Uses OpenAI GPT-4o to generate contextual, authentic comments based on
post content, existing comments, and engagement metrics.

Usage:
    from auto_engagement.ai_comment_generator import AICommentGenerator
    
    generator = AICommentGenerator()
    comment = generator.generate_comment(
        platform='instagram',
        post_content='Amazing sunset photo!',
        existing_comments=['So beautiful!', 'Where is this?'],
        username='photographer123',
        engagement='1.2K likes'
    )
"""

import os
import base64
import requests
from typing import List, Optional
from dataclasses import dataclass, field


@dataclass
class PostContext:
    """Context data for generating a comment."""
    platform: str
    username: str = ""
    post_content: str = ""
    visual_summary: str = ""
    existing_comments: List[str] = field(default_factory=list)
    engagement: str = ""
    post_url: str = ""


@dataclass
class GeneratedComment:
    """Result of comment generation."""
    text: str
    success: bool
    error: str = ""


class AICommentGenerator:
    """
    AI-powered comment generator using OpenAI GPT-4o.
    
    Generates authentic, contextual comments that:
    - Reference specific content from the post
    - Add to the conversation naturally
    - Feel human and relatable
    - Use appropriate emojis for each platform
    """
    
    PLATFORM_VIBES = {
        'instagram': 'supportive and engaging',
        'threads': 'conversational and thoughtful',
        'tiktok': 'casual and fun',
        'twitter': 'witty and concise',
        'youtube': 'appreciative and engaging'
    }
    
    def __init__(self, api_key: Optional[str] = None):
        """
        Initialize the comment generator.
        
        Args:
            api_key: OpenAI API key (defaults to OPENAI_API_KEY env var)
        """
        self.api_key = api_key or os.environ.get('OPENAI_API_KEY')
        if not self.api_key:
            raise ValueError("OpenAI API key required. Set OPENAI_API_KEY env var or pass api_key.")
    
    def generate_comment(
        self,
        platform: str,
        post_content: str,
        existing_comments: List[str] = None,
        username: str = "",
        engagement: str = "",
        max_length: int = 80
    ) -> GeneratedComment:
        """
        Generate a contextual comment for a social media post.
        
        Args:
            platform: Social platform (instagram, threads, tiktok, etc.)
            post_content: Main post content/caption/description
            existing_comments: List of existing comments on the post
            username: Creator's username
            engagement: Engagement stats string (e.g., "1.2K likes, 50 comments")
            max_length: Maximum comment length in characters
            
        Returns:
            GeneratedComment with text and success status
        """
        try:
            comments_text = "\n".join(existing_comments[:5]) if existing_comments else "No comments yet"
            platform_vibe = self.PLATFORM_VIBES.get(platform.lower(), 'engaging')
            
            prompt = f"""You are commenting on a {platform} post. Generate a SHORT, authentic comment (max {max_length} chars) with 1-2 emojis.

POST BY @{username}:
{post_content[:400]}

{f'ENGAGEMENT: {engagement}' if engagement else ''}

WHAT OTHERS ARE SAYING:
{comments_text}

Generate a thoughtful comment that:
- References specific content from the post when possible
- Adds to the conversation naturally (not just "great post!")
- Feels authentic and human
- Uses appropriate emojis for {platform}
- Matches the platform vibe: {platform_vibe}

Output ONLY the comment text:"""

            headers = {
                'Authorization': f'Bearer {self.api_key}',
                'Content-Type': 'application/json'
            }
            
            payload = {
                'model': 'gpt-4o',
                'messages': [{'role': 'user', 'content': prompt}],
                'max_tokens': 60,
                'temperature': 0.85
            }
            
            response = requests.post(
                'https://api.openai.com/v1/chat/completions',
                headers=headers,
                json=payload,
                timeout=20
            )
            
            if response.status_code == 200:
                text = response.json()['choices'][0]['message']['content'].strip().strip('"')
                return GeneratedComment(text=text, success=True)
            else:
                return GeneratedComment(
                    text="",
                    success=False,
                    error=f"API error: {response.status_code}"
                )
                
        except Exception as e:
            return GeneratedComment(text="", success=False, error=str(e))
    
    def analyze_image(self, image_path: str, prompt: str = None) -> str:
        """
        Analyze an image using OpenAI Vision.
        
        Args:
            image_path: Path to image file
            prompt: Analysis prompt (defaults to content description)
            
        Returns:
            Analysis text or empty string on failure
        """
        if not os.path.exists(image_path):
            return ""
        
        try:
            with open(image_path, 'rb') as f:
                image_data = base64.b64encode(f.read()).decode('utf-8')
            
            analysis_prompt = prompt or "Describe this social media post in one sentence - what's shown and the vibe."
            
            headers = {
                'Authorization': f'Bearer {self.api_key}',
                'Content-Type': 'application/json'
            }
            
            payload = {
                'model': 'gpt-4o',
                'messages': [{
                    'role': 'user',
                    'content': [
                        {'type': 'text', 'text': analysis_prompt},
                        {'type': 'image_url', 'image_url': {'url': f'data:image/png;base64,{image_data}'}}
                    ]
                }],
                'max_tokens': 150
            }
            
            response = requests.post(
                'https://api.openai.com/v1/chat/completions',
                headers=headers,
                json=payload,
                timeout=30
            )
            
            if response.status_code == 200:
                return response.json()['choices'][0]['message']['content']
                
        except Exception as e:
            print(f"Vision API error: {e}")
        
        return ""
    
    def generate_from_context(self, context: PostContext) -> GeneratedComment:
        """
        Generate comment from a PostContext object.
        
        Args:
            context: PostContext with all available data
            
        Returns:
            GeneratedComment result
        """
        content = context.post_content or context.visual_summary or ""
        return self.generate_comment(
            platform=context.platform,
            post_content=content,
            existing_comments=context.existing_comments,
            username=context.username,
            engagement=context.engagement
        )

File: packages/threads-comments/src/automation/ai-comment-generator.ts
/**
 * AI Comment Generator for Threads
 * 
 * Based on python/utils/ai_comment_generator.py
 * 
 * Analyzes post content, images, and existing comments
 * to generate contextual, engaging AI-powered comments.
 */

// PostContext from python/utils/ai_comment_generator.py
export interface PostContext {
  platform: string;
  username: string;
  postContent: string;
  visualSummary?: string;  // Description of images/videos
  existingComments: string[];
  engagement?: string;  // e.g., "1.2K likes, 50 comments"
  postUrl?: string;
  likeCount?: string;
  replyCount?: string;
}

export interface PostAnalysis {
  mainPost: string;
  username: string;
  replies: string[];
  hasImage: boolean;
  hasVideo: boolean;
  sentiment: 'positive' | 'negative' | 'neutral' | 'question';
  topics: string[];
  tone: string;
  engagement?: string;
  isInappropriate?: boolean;
  skipReason?: string;
}

// Content filter for inappropriate/thirst trap posts
const INAPPROPRIATE_KEYWORDS = [
  // Suggestive/thirst trap indicators
  'onlyfans', 'of link', 'link in bio', 'dm for more', 'dm me',
  'swipe up', 'exclusive content', 'subscribe', 'spicy', 'uncensored',
  'nsfw', '18+', 'adults only', 'mature content',
  // Body-focused thirst trap phrases
  'rate me', 'am i pretty', 'am i hot', 'do you like', 'what would you do',
  'smash or pass', 'would you date', 'sliding into', 'hit me up',
  // Spam/scam indicators
  'crypto', 'nft drop', 'free money', 'giveaway', 'dm to win',
  'make money fast', 'passive income', 'get rich',
];

const INAPPROPRIATE_EMOJIS = [
  '🍑', '🍆', '🥵', '💦', '🔞', '👅', '💋', '🤤',
];

export function isInappropriateContent(text: string): { inappropriate: boolean; reason?: string } {
  const lowerText = text.toLowerCase();
  
  // Check keywords
  for (const keyword of INAPPROPRIATE_KEYWORDS) {
    if (lowerText.includes(keyword)) {
      return { inappropriate: true, reason: `Contains "${keyword}"` };
    }
  }
  
  // Check emojis (multiple suggestive emojis = likely thirst trap)
  let emojiCount = 0;
  for (const emoji of INAPPROPRIATE_EMOJIS) {
    if (text.includes(emoji)) {
      emojiCount++;
    }
  }
  if (emojiCount >= 2) {
    return { inappropriate: true, reason: 'Multiple suggestive emojis detected' };
  }
  
  // Check for minimal text with just emojis (common thirst trap pattern)
  const textWithoutEmojis = text.replace(/[\u{1F300}-\u{1F9FF}]/gu, '').trim();
  if (textWithoutEmojis.length < 10 && text.length > 5) {
    // Very short text with emojis - check if it's just attention seeking
    const attentionPhrases = ['hey', 'hi', 'hello', 'look', 'check', 'new', 'just'];
    if (attentionPhrases.some(p => lowerText.includes(p)) && text.includes('📸')) {
      return { inappropriate: true, reason: 'Minimal text with photo indicator' };
    }
  }
  
  return { inappropriate: false };
}

export interface GeneratedComment {
  text: string;
  success: boolean;
  error?: string;
}

export interface AICommentConfig {
  provider: 'openai' | 'anthropic' | 'local';
  apiKey?: string;
  model?: string;
  maxLength?: number;
  temperature?: number;
  style?: 'engaging' | 'supportive' | 'insightful' | 'curious' | 'relatable';
}

// Platform vibes from python/utils/ai_comment_generator.py
const PLATFORM_VIBES: Record<string, string> = {
  instagram: 'supportive and engaging',
  threads: 'conversational and thoughtful',
  tiktok: 'casual and fun',
  twitter: 'witty and concise',
  youtube: 'appreciative and engaging',
};

const DEFAULT_CONFIG: AICommentConfig = {
  provider: 'openai',
  model: 'gpt-4o',
  maxLength: 80,
  temperature: 0.85,
  style: 'engaging',
};

export class ThreadsAICommentGenerator {
  private config: AICommentConfig;

  constructor(config: Partial<AICommentConfig> = {}) {
    // Load API key from environment if not provided
    const apiKey = config.apiKey || process.env.OPENAI_API_KEY;
    this.config = { 
      ...DEFAULT_CONFIG, 
      ...config,
      apiKey,
      provider: apiKey ? 'openai' : 'local',
    };
    
    if (apiKey) {
      console.log('[AI] ✅ OpenAI API key loaded - using GPT-4o');
    } else {
      console.log('[AI] ⚠️ No API key - using local templates');
    }
  }

  /**
   * Analyze post content to extract context
   */
  analyzePost(context: { mainPost: string; username: string; replies: string[] }): PostAnalysis {
    const { mainPost, username, replies } = context;
    
    // Detect sentiment
    const positiveWords = ['love', 'amazing', 'great', 'awesome', 'beautiful', 'happy', '😍', '🔥', '✨', '💯'];
    const negativeWords = ['hate', 'bad', 'terrible', 'sad', 'awful', 'annoying', '😢', '😡'];
    const questionWords = ['?', 'why', 'how', 'what', 'when', 'who', 'does', 'can'];
    
    const lowerPost = mainPost.toLowerCase();
    let sentiment: PostAnalysis['sentiment'] = 'neutral';
    
    if (questionWords.some(w => lowerPost.includes(w))) {
      sentiment = 'question';
    } else if (positiveWords.some(w => lowerPost.includes(w))) {
      sentiment = 'positive';
    } else if (negativeWords.some(w => lowerPost.includes(w))) {
      sentiment = 'negative';
    }
    
    // Detect topics from keywords
    const topicKeywords: Record<string, string[]> = {
      tech: ['code', 'software', 'ai', 'app', 'developer', 'programming', 'tech'],
      art: ['art', 'design', 'creative', 'draw', 'paint', 'artist', 'visual'],
      fitness: ['gym', 'workout', 'fitness', 'health', 'exercise', 'lift'],
      business: ['business', 'entrepreneur', 'startup', 'money', 'invest'],
      lifestyle: ['life', 'living', 'daily', 'routine', 'vibe', 'mood'],
      motivation: ['motivation', 'inspire', 'dream', 'goal', 'success', 'mindset'],
      food: ['food', 'eat', 'cook', 'recipe', 'restaurant', 'delicious'],
      travel: ['travel', 'trip', 'vacation', 'explore', 'adventure'],
      humor: ['lol', 'funny', 'joke', 'haha', '😂', '🤣', 'meme'],
    };
    
    const topics: string[] = [];
    for (const [topic, keywords] of Object.entries(topicKeywords)) {
      if (keywords.some(kw => lowerPost.includes(kw))) {
        topics.push(topic);
      }
    }
    
    // Detect media type from context
    const hasImage = lowerPost.includes('view') || lowerPost.includes('image') || lowerPost.includes('photo');
    const hasVideo = lowerPost.includes('video') || lowerPost.includes('watch');
    
    // Detect tone
    const tones = {
      casual: ['lol', 'lmao', 'gonna', 'wanna', 'ya', 'yall'],
      professional: ['therefore', 'however', 'regarding', 'professional'],
      emotional: ['!', '❤️', '💔', 'feel', 'feeling', 'love', 'hate'],
    };
    
    let tone = 'neutral';
    for (const [t, words] of Object.entries(tones)) {
      if (words.some(w => lowerPost.includes(w))) {
        tone = t;
        break;
      }
    }
    
    // Check for inappropriate content (thirst traps, spam, etc.)
    const contentCheck = isInappropriateContent(mainPost);
    
    return {
      mainPost,
      username,
      replies,
      hasImage,
      hasVideo,
      sentiment,
      topics: topics.length > 0 ? topics : ['general'],
      tone,
      isInappropriate: contentCheck.inappropriate,
      skipReason: contentCheck.reason,
    };
  }

  /**
   * Generate an AI-powered comment based on post analysis
   * Based on python/utils/ai_comment_generator.py generate_comment()
   */
  async generateComment(analysis: PostAnalysis): Promise<string> {
    if (this.config.provider === 'local' || !this.config.apiKey) {
      return this.generateLocalComment(analysis);
    }
    
    // Use OpenAI or Anthropic for AI generation
    const prompt = this.buildPrompt(analysis);
    
    if (this.config.provider === 'openai') {
      return this.generateWithOpenAI(prompt);
    } else if (this.config.provider === 'anthropic') {
      return this.generateWithAnthropic(prompt);
    }
    
    // Fallback to local
    return this.generateLocalComment(analysis);
  }

  /**
   * Generate comment from full PostContext (like Python's generate_from_context)
   */
  async generateFromContext(context: PostContext): Promise<GeneratedComment> {
    try {
      const analysis = this.analyzePost({
        mainPost: context.postContent || context.visualSummary || '',
        username: context.username,
        replies: context.existingComments,
      });
      
      // Add engagement info
      if (context.engagement) {
        analysis.engagement = context.engagement;
      }
      
      const text = await this.generateComment(analysis);
      return { text, success: true };
    } catch (error) {
      return { text: '', success: false, error: String(error) };
    }
  }

  /**
   * Build prompt matching python/utils/ai_comment_generator.py
   */
  private buildPrompt(analysis: PostAnalysis): string {
    const { mainPost, username, replies, sentiment, topics, tone, engagement } = analysis;
    const platformVibe = PLATFORM_VIBES.threads;
    const commentsText = replies.length > 0 
      ? replies.slice(0, 5).map(r => r.substring(0, 100)).join('\n')
      : 'No comments yet';
    
    // Prompt structure from Python script
    const prompt = `You are commenting on a Threads post. Generate a SHORT, authentic comment (max ${this.config.maxLength} chars) with 1-2 emojis.

POST BY @${username}:
${mainPost.substring(0, 400)}

${engagement ? `ENGAGEMENT: ${engagement}` : ''}

WHAT OTHERS ARE SAYING:
${commentsText}

ANALYSIS:
- Sentiment: ${sentiment}
- Topics: ${topics.join(', ')}
- Tone: ${tone}

Generate a thoughtful comment that:
- References specific content from the post when possible
- Adds to the conversation naturally (not just "great post!")
- Feels authentic and human
- Uses appropriate emojis for Threads
- Matches the platform vibe: ${platformVibe}
${sentiment === 'question' ? '- Consider answering or engaging with the question' : ''}

Output ONLY the comment text:`;

    return prompt;
  }

  private async generateWithOpenAI(prompt: string): Promise<string> {
    const fallback = { mainPost: '', username: '', replies: [], hasImage: false, hasVideo: false, sentiment: 'neutral' as const, topics: ['general'], tone: 'neutral' };
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 15000);
    try {
      const response = await fetch('https://api.openai.com/v1/chat/completions', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.config.apiKey}`,
        },
        body: JSON.stringify({
          model: this.config.model,
          messages: [
            {
              role: 'system',
              content: 'You are a social media engagement expert. Generate authentic, contextual comments that sound natural. Never be generic or spammy.',
            },
            { role: 'user', content: prompt },
          ],
          max_tokens: 100,
          temperature: this.config.temperature,
        }),
        signal: controller.signal,
      });
      clearTimeout(timeout);

      if (!response.ok) {
        console.error('OpenAI error, falling back to local');
        return this.generateLocalComment(fallback);
      }

      const data = await response.json() as { choices?: { message?: { content?: string } }[] };
      return data.choices?.[0]?.message?.content?.trim() || this.generateLocalComment(fallback);
    } catch (error) {
      clearTimeout(timeout);
      console.error('OpenAI request failed:', error instanceof Error ? error.message : error);
      return this.generateLocalComment(fallback);
    }
  }

  private async generateWithAnthropic(prompt: string): Promise<string> {
    const fallback = { mainPost: '', username: '', replies: [], hasImage: false, hasVideo: false, sentiment: 'neutral' as const, topics: ['general'], tone: 'neutral' };
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 15000);
    try {
      const response = await fetch('https://api.anthropic.com/v1/messages', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'x-api-key': this.config.apiKey!,
          'anthropic-version': '2023-06-01',
        },
        body: JSON.stringify({
          model: this.config.model ?? 'claude-3-haiku-20240307',
          max_tokens: 100,
          messages: [{ role: 'user', content: prompt }],
        }),
        signal: controller.signal,
      });
      clearTimeout(timeout);

      if (!response.ok) {
        console.error('Anthropic error, falling back to local');
        return this.generateLocalComment(fallback);
      }

      const data = await response.json() as { content?: { text?: string }[] };
      return data.content?.[0]?.text?.trim() || this.generateLocalComment(fallback);
    } catch (error) {
      clearTimeout(timeout);
      console.error('Anthropic request failed:', error instanceof Error ? error.message : error);
      return this.generateLocalComment(fallback);
    }
  }

  /**
   * Generate comment locally based on analysis (no API needed)
   */
  private generateLocalComment(analysis: PostAnalysis): string {
    const { sentiment, topics, tone } = analysis;
    
    // Topic-specific templates
    const topicTemplates: Record<string, string[]> = {
      tech: [
        "This is exactly the kind of innovation we need 🔥",
        "The tech behind this is impressive! Building something similar?",
        "This changes the game for developers 💻",
        "Would love to see more breakdowns like this ⚡",
      ],
      art: [
        "The creativity here is unmatched ✨",
        "Your artistic vision is incredible 🎨",
        "This is the kind of art that stops you scrolling",
        "The detail in this is amazing 👏",
      ],
      fitness: [
        "This is the motivation I needed today 💪",
        "Consistency really is everything 🏋️",
        "Adding this to my routine immediately",
        "The discipline here is inspiring 🔥",
      ],
      business: [
        "This perspective on growth is spot on 📈",
        "More founders need to hear this 💯",
        "The business insight here is valuable",
        "This is the kind of strategic thinking that wins",
      ],
      motivation: [
        "Needed to hear this today 🙌",
        "This hit different. Saving this one ⭐",
        "The mindset shift this creates is powerful",
        "Printing this and putting it on my wall 📌",
      ],
      humor: [
        "I literally cannot stop laughing at this 😂",
        "Why is this so accurate though 💀",
        "I feel personally called out 😅",
        "This just made my whole day 🤣",
      ],
      lifestyle: [
        "This is the vibe we're all chasing ✨",
        "Living life right! This is goals 🙌",
        "The energy in this is immaculate",
        "This is what it's all about 💯",
      ],
      general: [
        "This really resonates with me ✨",
        "Appreciate you sharing this perspective 🙌",
        "More of this content please! 🔥",
        "This is what my feed needed today 💯",
      ],
    };
    
    // Sentiment-based adjustments
    const sentimentPrefixes: Record<string, string[]> = {
      question: [
        "Great question! ",
        "I think about this too - ",
        "My take: ",
      ],
      positive: [
        "",
        "Yes! ",
        "100%! ",
      ],
      negative: [
        "I feel this. ",
        "Been there. ",
        "This is real. ",
      ],
      neutral: [
        "",
        "Interesting - ",
        "",
      ],
    };
    
    // Select templates based on first topic
    const topic = topics[0] || 'general';
    const templates = topicTemplates[topic] || topicTemplates.general;
    const template = templates[Math.floor(Math.random() * templates.length)];
    
    // Add sentiment prefix sometimes
    const prefixes = sentimentPrefixes[sentiment] || sentimentPrefixes.neutral;
    const prefix = Math.random() > 0.7 ? prefixes[Math.floor(Math.random() * prefixes.length)] : '';
    
    return prefix + template;
  }
}
